mne_sandbox/connectivity/cfc.py

```diff
@@ -0,0 +1,506 @@
+    return data_ph, data_am, ix_map_ph, ix_map_am
+
+
+def _filter_ph_am(xph, xam, f_ph, f_am, sfreq, filterfn=None, kws_filt=None):
+    """Aux function for phase/amplitude filtering for one pair of channels"""
+    from ..externals.pacpy.pac import _range_sanity
+    from scipy.signal import hilbert
+    filterfn = band_pass_filter if filterfn is None else filterfn
+    kws_filt = {} if kws_filt is None else kws_filt
+
+    # Filter the two signals + hilbert/phase
+    _range_sanity(f_ph, f_am)
+    xph = filterfn(xph, sfreq, *f_ph)
+    xam = filterfn(xam, sfreq, *f_am)
+
+    xph = np.angle(hilbert(xph))
+    xam = np.abs(hilbert(xam))
+    return xph, xam
+
+
+def _array_raw_to_epochs(x, sfreq, ev, tmin, tmax):
+    """Aux function to create epochs from a 2D array"""
+    if ev.ndim != 1:
+        raise ValueError('ev must be 1D')
+    if ev.dtype != int:
+        raise ValueError('ev must be of dtype int')
+
+    # Check that events won't be cut off
+    n_times = x.shape[-1]
+    min_ix = 0 - sfreq * tmin
+    max_ix = n_times - sfreq * tmax
+    msk_keep = np.logical_and(ev > min_ix, ev < max_ix)
+
+    if not all(msk_keep):
+        print('Some events will be cut off!')
+        ev = ev[msk_keep]
+
+    # Pull events from the raw data
+    epochs = []
+    for ix in ev:
+        ix_min, ix_max = [ix + int(i_tlim * sfreq)
+                          for i_tlim in [tmin, tmax]]
+        epochs.append(x[np.newaxis, :, ix_min:ix_max])
+    epochs = np.concatenate(epochs, axis=0)
+    times = np.arange(epochs.shape[-1]) / float(sfreq) + tmin
+    return epochs, times, msk_keep
+
+
+# For the viz functions
+def _extract_phase_and_amp(data_phase, data_amp, sfreq, freqs_phase,
+                           freqs_amp, scale=True):
+    """Extract the phase and amplitude of two signals for PAC viz.
+    data should be shape (n_epochs, n_times)"""
+    from sklearn.preprocessing import scale
+
+    # Morlet transform to get complex representation
+    band_ph = cwt_morlet(data_phase, sfreq, freqs_phase)
+    band_amp = cwt_morlet(data_amp, sfreq, freqs_amp)
+
+    # Calculate the phase/amplitude of relevant signals across epochs
+    band_ph_stacked = np.hstack(np.real(band_ph))
+    angle_ph = np.hstack(np.angle(band_ph))
+    amp = np.hstack(np.abs(band_amp) ** 2)
+
+    # Scale the amplitude for viz so low freqs don't dominate highs
+    if scale is True:
+        amp = scale(amp, axis=1)
+    return angle_ph, band_ph_stacked, amp
+
+
+def _pull_data(inst, ix_ph, ix_amp, ev=None, tmin=None, tmax=None):
+    """Pull data from either Base or Epochs instances"""
+    from mne.io.base import _BaseRaw
+    from mne.epochs import _BaseEpochs
+    if isinstance(inst, _BaseEpochs):
+        data_ph = inst._data[:, ix_ph, :]
+        data_amp = inst._data[:, ix_amp, :]
+    elif isinstance(inst, _BaseRaw):
+        data = inst[[ix_ph, ix_amp], :][0].squeeze()
+        data_ph, data_amp = [i[np.newaxis, ...] for i in data]
```

Comment:
`ii` not `i`

Also, I don't get what's done here. You're just getting two vars, why do you use a loop?